parser/parser/cc/grammars/typedruby.ypp

```diff
@@ -2785,6 +2785,12 @@ opt_block_args_tail:
                         lambda->end);
                       DIAGCHECK();
                     }
+                | t_lambda error
+                    {
+                      driver.lex.unextend();
+                      driver.lex.context.inLambda = $1->context.inLambda;
```

Comment:
Why do we only take these two, and not the `numparam_stack.pop` nor the `cmdarg.pop` from the other case?

Reply:
`numparam_stack.pop()` should definitely be in there, thanks!

It's hard with `cmdarg`, as we don't know if the argument list parsed successfully. For example, if we change line 2 in the `crash_after_parse_errors.rb` test from `->e` to `-><`:

```
# typed: true
-><
r=e
r=e
r&&o
```

The production that parses the argument list won't be selected, and `cmdarg.push` will not happen. It's not obvious to me if it's possible to detect if that case happened, but maybe I'm missing something?

Reply:
Usually the way you have to deal with cases like this is handle the case where the argument list was parsed and then you saw an error token, and handle the case where the argument list was not parsed before seeing the error token.

Maybe a better way would be to have some sort of RAII-style production rule, so that if a node ever had its destructor run (because e.g. it was popped during the error recovery algorithm) then the destructor calls `.pop()`.

Short of that, I'm not sure.

Probably we have this problem with ~every other place that pushes and pops things from the lexer. I would not be surprised if we could make all sorts of bugs with knowledge of this.